cogito/protocol.go

```diff
@@ -257,7 +263,11 @@ func (src *Source) Validate() error {
 	//
 	// Validate optional fields.
 	//
-	// In this case, nothing to validate.
+	if src.GhHostname != "" {
+		if !hostnameRegexp.MatchString(src.GhHostname) {
+			return fmt.Errorf("source: invalid github_api_hostname: %s. Don't configure the schema or the path", src.GhHostname)
+		}
+	}
```

Comment:
Not a fan of your suggestion but if that is the team decision, I am ready to do the change.

My point of view:
- for me nested is more readable...
- I got hurt so many times with more complex if statements with multiple conditionals that I try to avoid it... 
 On the other hand, maybe in some simpler cases your approach is justified...

Reply:
I think that the reason of Olivier's suggestion is because:
- there is nothing inside the `if` besides the call to hostnameRegexp.MatchString.
- the expression of the `if` is protecting the call of hostnameRegexp.MatchString. A classic would be:
`if ptr != nil && foo(*ptr)`
it is indeed idiomatic to use the && in those cases.

> I got hurt so many times with more complex if statements with multiple conditionals that I try to avoid it...

Very true, this happens to everybody and in any case makes the code easier to misunderstand.

But here we are in the realm of the idiomatic `if ptr != nil && foo(*ptr)`.

Now, there is also another approach that might work in this case: see if we can apply the defaults before arriving here. If we can, then this code would become simply

```
if !hostnameRegexp.MatchString(src.GhHostname) 
```

even simpler and arguably even better.

Reply:
Yes, moving this statement after defaults are applied simplifies a lot and additionally verifies that someone doesn't put the wrong value for: `github.GhDefaultHostname`. So win win situation. Thanks to both.

```
if !hostnameRegexp.MatchString(src.GhHostname) {
		return fmt.Errorf("source: invalid github_api_hostname: %s. Don't configure the schema or the path", src.GhHostname)
}
```